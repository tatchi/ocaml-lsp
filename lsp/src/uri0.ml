open Import

module Private = struct
  let win32 = ref Sys.win32
end

type t = Uri_lexer.t =
  { scheme : string option
  ; authority : string
  ; path : string
  }

let t_of_yojson json = Json.Conv.string_of_yojson json |> Uri_lexer.of_string

let to_string { scheme; authority; path } =
  let b = Buffer.create 64 in
  scheme
  |> Option.iter (fun s ->
         Buffer.add_string b s;
         Buffer.add_char b ':');
  Buffer.add_string b "//";
  Buffer.add_string b authority;
  if not (String.is_prefix path ~prefix:"/") then Buffer.add_char b '/';
  Buffer.add_string b
    (path |> Uri.pct_encode
    |> String.replace_all ~pattern:"%2F" ~with_:"/"
    (* https://github.com/microsoft/vscode-uri/blob/96acdc0be5f9d5f2640e1c1f6733bbf51ec95177/src/uri.ts#L453 *)
    |> String.replace_all ~pattern:":" ~with_:"%3A"
    |> String.replace_all ~pattern:"?" ~with_:"%3F"
    |> String.replace_all ~pattern:"#" ~with_:"%23"
    |> String.replace_all ~pattern:"[" ~with_:"%5B"
    |> String.replace_all ~pattern:"]" ~with_:"%5D"
    |> String.replace_all ~pattern:"@" ~with_:"%40"
    |> String.replace_all ~pattern:"!" ~with_:"%21"
    |> String.replace_all ~pattern:"$" ~with_:"%24"
    |> String.replace_all ~pattern:"&" ~with_:"%26"
    |> String.replace_all ~pattern:"'" ~with_:"%27"
    |> String.replace_all ~pattern:"(" ~with_:"%28"
    |> String.replace_all ~pattern:")" ~with_:"%29"
    |> String.replace_all ~pattern:"*" ~with_:"%2A"
    |> String.replace_all ~pattern:"+" ~with_:"%2B"
    |> String.replace_all ~pattern:"," ~with_:"%2C"
    |> String.replace_all ~pattern:";" ~with_:"%3B"
    |> String.replace_all ~pattern:"=" ~with_:"%3D"
    |> String.replace_all ~pattern:" " ~with_:"%20");
  Buffer.contents b

let yojson_of_t t = `String (to_string t)

let equal = ( = )

let compare (x : t) (y : t) = Stdlib.compare x y

let hash = Hashtbl.hash

let to_dyn { scheme; authority; path } =
  let open Dyn in
  record
    [ ("scheme", (option string) scheme)
    ; ("authority", string authority)
    ; ("path", string path)
    ]

let to_path t =
  let path =
    t.path
    (* |> String.replace_all ~pattern:"\\" ~with_:"/" |> String.replace_all
       ~pattern:"%5C" ~with_:"/" |> String.replace_all ~pattern:"%3A" ~with_:":"
       |> String.replace_all ~pattern:"%20" ~with_:" " |> String.replace_all
       ~pattern:"%3D" ~with_:"=" |> String.replace_all ~pattern:"%3F"
       ~with_:"?" *)
  in
  path

let of_path (path : string) =
  let path = Uri_lexer.escape_path path in
  { path; scheme = Some "file"; authority = "" }
