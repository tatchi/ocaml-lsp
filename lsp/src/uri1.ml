open Import

module Private = struct
  let win32 = ref Sys.win32
end

type t =
  { scheme : string
  ; authority : string
  ; path : string
  }

let path t = t.path

let backslash_to_slash =
  String.map ~f:(function
    | '\\' -> '/'
    | _ as c -> c)

let slash_to_backslash =
  String.map ~f:(function
    | '/' -> '\\'
    | _ as c -> c)

let of_path path =
  let path = if !Private.win32 then backslash_to_slash path else path in
  let path, authority =
    let len = String.length path in
    if len = 0 then ("/", "") (* TODO: use String.is_prefix or start_with? *)
    else if len > 1 && path.[0] = '/' && path.[1] = '/' then (
      let idx = String.index_from_opt path 2 '/' in
      match idx with
      | None -> ("/", String.sub path ~pos:2 ~len:(len - 2))
      | Some i ->
        let authority = String.sub path ~pos:2 ~len:(i - 2) in
        let path =
          let path = String.sub path ~pos:i ~len:(len - i) in
          if path = "" then "/" else path
        in
        (path, authority))
    else (path, "")
  in
  let path = if path.[0] <> '/' then "/" ^ path else path in
  (* Printf.printf "scheme: %s - authority: %s - path: %s\n" "file" authority path; *)
  { scheme = "file"; authority; path }

let to_path { path; authority; scheme } =
  let path =
    let len = String.length path in
    if len = 0 then "/"
    else if authority <> "" && len > 1 && scheme = "file" then
      "//" ^ authority ^ path
    else if len < 3 then path
    else
      let c0 = path.[0] in
      let c1 = path.[1] in
      let c2 = path.[2] in
      if
        c0 = '/'
        && ((c1 >= 'A' && c1 <= 'Z') || (c1 >= 'a' && c1 <= 'z'))
        && c2 = ':'
      then
        String.make 1 (Char.lowercase_ascii c1)
        ^ String.sub path ~pos:2 ~len:(String.length path - 2)
      else path
  in
  if !Private.win32 then slash_to_backslash path else path

let of_string s =
  let re =
    Re.Perl.re "^(([^:/?#]+?):)?(\\/\\/([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?"
    |> Re.compile
  in
  let res = Re.exec re s in
  let group re n = Re.Group.get_opt re n |> Option.value ~default:"" in
  let scheme = group res 2 in
  let authority = group res 4 in
  let path = group res 5 |> Uri.pct_decode in
  { scheme; authority; path }
