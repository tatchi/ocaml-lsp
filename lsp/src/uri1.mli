type t

val of_path : string -> t

val to_path : t -> string

val path : t -> string

val of_string : string -> t

val to_string : t -> string

module Private : sig
  val win32 : bool ref
end
