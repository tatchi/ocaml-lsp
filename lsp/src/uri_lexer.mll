{
type t =
  { scheme : string option
  ; authority : string
  ; path : string
  }
}

rule path buff= parse
| '/'? (['A'-'Z']|['a'-'z'] as driver_letter) ':' 
      { Buffer.add_char buff (Char.lowercase_ascii driver_letter); Buffer.add_char buff ':'; path1 buff lexbuf }
| [^ '/'] as c { Buffer.add_char buff '/'; Buffer.add_char buff c; path1 buff lexbuf }
| _ as c { Buffer.add_char buff c ; path1 buff lexbuf }
| eof { Buffer.add_char buff '/' ; Buffer.contents buff }
and path1 buf = parse
(* | '\\' { Buffer.add_char buf '/' ; path1 buf lexbuf } *)
(* | "%5" ['c' 'C'] { Buffer.add_char buf '/' ; path1 buf lexbuf }
| "%3" ['a' 'A'] { Buffer.add_char buf ':' ; path1 buf lexbuf }
| "%3" ['d' 'D'] { Buffer.add_char buf '=' ; path1 buf lexbuf }
| "%3" ['f' 'F'] { Buffer.add_char buf '?' ; path1 buf lexbuf }
| "%20" { Buffer.add_char buf ' ' ; path1 buf lexbuf } *)
| _ as c { Buffer.add_char buf c ; path1 buf lexbuf }
| eof { Buffer.contents buf }

and uri buff = parse
| ([^ ':']+) as scheme ':' { uri1 buff (Some scheme) lexbuf }
| "" { uri1 buff None lexbuf }
and uri1 buff scheme = parse
| "//" ([^ '/']* as authority) { uri2 buff scheme authority lexbuf }
| "" { uri2 buff scheme "" lexbuf }
and uri2 buff scheme authority = parse
| "" { { scheme ; authority ; path = path buff lexbuf } }

{
  let escape_path s =
    let lexbuf = Lexing.from_string s in
    path (Buffer.create 12) lexbuf

  let of_string s =
    let lexbuf = Lexing.from_string s in
    uri (Buffer.create 12) lexbuf
}
