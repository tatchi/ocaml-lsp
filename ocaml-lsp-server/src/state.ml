open Import

type init =
  | Uninitialized
  | Initialized of InitializeParams.t

type t =
  { store : Document_store.t
  ; merlin : Scheduler.thread
  ; init : init
  ; scheduler : Scheduler.t
  ; detached : Fiber.Pool.t
  ; configuration : Configuration.t
  ; trace : TraceValue.t
  ; ocamlformat : Fmt.t
  }
