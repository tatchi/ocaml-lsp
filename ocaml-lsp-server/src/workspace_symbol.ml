open Import

let to_loc (loc : Warnings.loc) : Loc.t =
  { loc_start = loc.loc_start
  ; loc_end = loc.loc_end
  ; loc_ghost = loc.loc_ghost
  }

let cmt_sign (cmt_infos : Cmt_format.cmt_infos) =
  match cmt_infos.cmt_annots with
  | Cmt_format.Implementation { Typedtree.str_type = sign; _ }
  | Cmt_format.Interface { Typedtree.sig_type = sign; _ }
  | Cmt_format.Packed (sign, _) ->
    sign
  | _ -> []

let loc_of_sig_item (signature_item : Types.signature_item) =
  match signature_item with
  | Types.Sig_value (_, descr, _) -> descr.val_loc
  | Types.Sig_type (_, descr, _, _) -> descr.type_loc
  | Types.Sig_typext (_, descr, _, _) -> descr.ext_loc
  | Types.Sig_module (_, _, descr, _, _) -> descr.md_loc
  | Types.Sig_modtype (_, descr, _) -> descr.mtd_loc
  | Types.Sig_class (_, descr, _, _) -> descr.cty_loc
  | Types.Sig_class_type (_, descr, _, _) -> descr.clty_loc

let id_of_sig_item (signature_item : Types.signature_item) =
  match signature_item with
  | Types.Sig_value (id, _, _)
  | Types.Sig_type (id, _, _, _)
  | Types.Sig_typext (id, _, _, _)
  | Types.Sig_module (id, _, _, _, _)
  | Types.Sig_modtype (id, _, _)
  | Types.Sig_class (id, _, _, _)
  | Types.Sig_class_type (id, _, _, _) ->
    id

let kind_of_sig_item (signature_item : Types.signature_item) : SymbolKind.t =
  match signature_item with
  | Types.Sig_value _ -> Function
  | Types.Sig_type _ -> String
  (* | Types.Sig_typext (_, _, Types.Text_exception, _) -> Constructor *)
  | Types.Sig_typext _ -> Constructor
  | Types.Sig_module _ -> Module
  | Types.Sig_modtype _ -> Module
  | Types.Sig_class _ -> Class
  | Types.Sig_class_type _ -> Class

let symbol_of_signature_item ~sourcefile (signature_item : Types.signature_item)
    =
  let id = id_of_sig_item signature_item in
  let loc = loc_of_sig_item signature_item in
  let name = Ident.name id in
  let kind = kind_of_sig_item signature_item in
  let range = Range.of_loc (loc |> to_loc) in
  let uri = sourcefile |> Uri.of_path in
  let location = Location.create ~uri ~range in
  SymbolInformation.create ~name ~kind ~location ()

let build_roots =
  (* by increasing order of priority *)
  [ "_darcs"; ".hg"; ".git"; "jengaroot.ml"; "omakeroot"; "_build"; "_obuild" ]

let find_build_dir path =
  let ( / ) = Filename.concat in
  let files = Sys.readdir path in
  let _, root =
    let rec memsuffix x = function
      | a :: r ->
        if x = a then
          Some r
        else
          memsuffix x r
      | [] -> None
    in
    Array.fold_left
      ~f:(fun (roots, found) f ->
        match memsuffix f roots with
        | None -> (roots, found)
        | Some roots -> (roots, Some f))
      ~init:(build_roots, None) files
  in
  match root with
  | None -> None
  | Some (("_obuild" | "_build") as dir) -> Some (path / dir)
  | Some _ -> Some path

let project_root ?(path = Sys.getcwd ()) () =
  let ( / ) = Filename.concat in
  let home =
    try Sys.getenv "HOME" with
    | Not_found -> ""
  in
  let path =
    if Filename.is_relative path then
      Sys.getcwd () / path
    else
      path
  in
  let rec find path =
    if path = home then
      None
    else
      match find_build_dir path with
      | None ->
        let parent = Filename.dirname path in
        if path = parent then
          None
        else
          find parent
      | Some build -> Some (path, build)
  in
  match find path with
  | None -> (None, None)
  | Some (root, build) -> (Some root, Some build)

let unique_subdirs ?(skip = fun _ -> false) dir_list =
  let rec subdirs acc path =
    Array.fold_left
      ~f:(fun acc p ->
        if skip p then
          acc
        else
          let path = Filename.concat path p in
          if
            try Sys.is_directory path with
            | Sys_error _ -> false
          then
            subdirs acc path
          else
            acc)
      ~init:(path :: acc) (Sys.readdir path)
  in
  let remove_dups l =
    let rec aux = function
      | a :: (b :: _ as r) when a = b -> aux r
      | a :: r -> a :: aux r
      | [] -> []
    in
    aux (List.sort ~compare:(fun a b -> Ordering.of_int (compare a b)) l)
  in
  remove_dups (List.fold_left ~f:subdirs ~init:[] dir_list)

let read_cmt cmt_file =
  let filename_from_path (path : string) =
    let parts = String.split_on_char ~sep:'/' path in
    List.last parts |> Option.value ~default:""
  in
  let cmt = Cmt_format.read_cmt cmt_file in
  let sourcefile =
    Sys.getcwd () ^ "/" ^ Option.value ~default:"" cmt.cmt_sourcefile
  in
  let signatures = cmt_sign cmt in
  signatures
  |> List.fold_left
       ~f:(fun acc signature_item ->
         let loc = loc_of_sig_item signature_item in
         let pos_filename = filename_from_path loc.loc_start.pos_fname in
         let filename = filename_from_path sourcefile in
         if pos_filename = filename then
           symbol_of_signature_item ~sourcefile signature_item :: acc
         else
           acc)
       ~init:[]

type cm_file =
  | Cmt of string
  | Cmti of string

let string_of_cm cm =
  match cm with
  | Cmt f
  | Cmti f ->
    f

module Cm_Files = Map.Make (struct
  type t = string

  let compare a b = Stdlib.compare a b |> Ordering.of_int

  let to_dyn t = Dyn.String t
end)

let find_dirfiles dirs =
  let choose_file f1 f2 =
    match (f1, f2) with
    | (Cmt _ as f), _
    | _, (Cmt _ as f) ->
      f
    | (Cmti _ as f), Cmti _ -> f
  in
  let split_filename file =
    let maybe_index = String.rindex_opt file '.' in
    let len = String.length file in
    match maybe_index with
    | None -> (file, "")
    | Some index ->
      let modul = String.capitalize_ascii (String.sub file ~pos:0 ~len:index) in
      let ext =
        String.lowercase_ascii
          (String.sub file ~pos:(index + 1) ~len:(len - index - 1))
      in
      (modul, ext)
  in
  List.fold_left
    ~f:(fun files dir ->
      Sys.readdir dir
      |> Array.fold_left
           ~f:(fun acc file ->
             let modul, ext = split_filename file in
             let path = dir ^ "/" ^ file in
             match ext with
             | "cmt" -> Cm_Files.set acc modul (Cmt path)
             | "cmti" -> (
               let current_file = Cm_Files.find acc modul in
               let cmti_file = Cmti path in
               match current_file with
               | None -> Cm_Files.set acc modul cmti_file
               | Some current_file ->
                 Cm_Files.set acc modul (choose_file current_file cmti_file))
             | _ -> acc)
           ~init:files)
    ~init:Cm_Files.empty dirs
  |> Cm_Files.values |> List.map ~f:string_of_cm

let run ({ query; _ } : WorkspaceSymbolParams.t) (rootUri : Uri.t option) =
  let rootUri =
    Option.map rootUri ~f:Uri.to_path |> Option.value ~default:"No root uri"
  in
  let _, build_dir = project_root ~path:rootUri () in
  let build_dir = Option.value build_dir ~default:"No build dir" in
  let unique_subdirs = unique_subdirs [ build_dir ] in
  let all_files = find_dirfiles unique_subdirs in
  let load_file file =
    let maybe_index = String.rindex_opt file '.' in
    match maybe_index with
    | None -> []
    | Some i -> (
      let len = String.length file in
      let ext =
        String.lowercase_ascii (String.sub file ~pos:(i + 1) ~len:(len - i - 1))
      in
      match ext with
      | "cmt"
      | "cmti" ->
        read_cmt file
      | _ -> [])
  in
  let all_symbols = List.map ~f:load_file all_files |> List.flatten in

  match query with
  | "" -> all_symbols
  | query ->
    let reg_exp = Str.regexp_string query in
    all_symbols
    |> List.filter ~f:(fun symbol ->
           try
             let () =
               ignore
                 (Str.search_forward reg_exp SymbolInformation.(symbol.name) 0)
             in
             true
           with
           | Not_found -> false)
