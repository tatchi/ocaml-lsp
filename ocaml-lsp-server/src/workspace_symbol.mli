open Import

val run : WorkspaceSymbolParams.t -> Uri.t option -> SymbolInformation.t list
