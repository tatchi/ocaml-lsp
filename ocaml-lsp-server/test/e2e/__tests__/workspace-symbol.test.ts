import * as path from "path";
import * as child_process from "child_process";
import * as Types from "vscode-languageserver-types";
import * as LanguageServer from "./../src/LanguageServer";

describe("workspace/symbol", () => {
  let languageServer = null;

  let testWorkspacePath = path.join(__dirname, "workspace_symbol/");

  beforeEach(async () => {
    languageServer = await LanguageServer.startAndInitialize({
      rootUri: LanguageServer.toURI(testWorkspacePath),
    });
  });

  afterEach(async () => {
    await LanguageServer.exit(languageServer);
    languageServer = null;
  });

  async function queryWorkspaceSymbol(query: string) {
    return await languageServer.sendRequest("workspace/symbol", {
      query,
    });
  }

  it("returns symbols from workspace", async () => {
    child_process.execSync("dune build", { cwd: testWorkspacePath });

    let result: Types.SymbolInformation[] = await queryWorkspaceSymbol("");

    expect(
      result.map((symbolInfo) => ({
        ...symbolInfo,
        location: {
          ...symbolInfo.location,
          uri: symbolInfo.location.uri.replace(testWorkspacePath, '<testWorkspacePath>/'),
        },
      })),
    ).toMatchSnapshot();
  });
});
