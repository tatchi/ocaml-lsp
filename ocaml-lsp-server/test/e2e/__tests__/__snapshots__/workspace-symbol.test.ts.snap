// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workspace/symbol returns symbols from workspace 1`] = `
Array [
  Object {
    "kind": 2,
    "location": Object {
      "range": Object {
        "end": Object {
          "character": 36,
          "line": 5,
        },
        "start": Object {
          "character": 0,
          "line": 5,
        },
      },
      "uri": "file://<testWorkspacePath>/workspace_symbol.ml-gen",
    },
    "name": "Main",
  },
  Object {
    "kind": 2,
    "location": Object {
      "range": Object {
        "end": Object {
          "character": 34,
          "line": 1,
        },
        "start": Object {
          "character": 0,
          "line": 1,
        },
      },
      "uri": "file://<testWorkspacePath>/workspace_symbol.ml-gen",
    },
    "name": "Lib",
  },
  Object {
    "kind": 12,
    "location": Object {
      "range": Object {
        "end": Object {
          "character": 5,
          "line": 0,
        },
        "start": Object {
          "character": 4,
          "line": 0,
        },
      },
      "uri": "file://<testWorkspacePath>/lib.ml",
    },
    "name": "x",
  },
  Object {
    "kind": 12,
    "location": Object {
      "range": Object {
        "end": Object {
          "character": 5,
          "line": 0,
        },
        "start": Object {
          "character": 4,
          "line": 0,
        },
      },
      "uri": "file://<testWorkspacePath>/main.ml",
    },
    "name": "y",
  },
]
`;
